### DISCUSSION TITLE
Central repository for public replacement data

### DISCUSSION THREAD

**Discussion by raymcdermott** [repl-acement/repl-acement#23] [Ideas]
Create support for a central immutable store of data that is imported to replacement

**Comment by andrewzhurov** [comment]
Would it be akin to https://github.com/clojars/clojars-web ?
If so, who's to be the authority?

**Comment by andrewzhurov** [comment]
Why?

**Reply by andrewzhurov** [reply]
To share code

**Reply by andrewzhurov** [reply]
To discover code

**Reply by andrewzhurov** [reply]
To discover common code (in order to save trouble of indexing it for end-users)

**Comment by andrewzhurov** [comment]
Hosting of content-based addressable codebases would need to be snappy, akin to git push.
Having just the central repo would mean end-users would need to be authorised to write it.
But users may have malicious intent in pushing spam/junk, and as such repo is immutable, there would be need to prevent that for stopping it from bloating up, as I presume [this discussion is about](https://github.com/repl-acement/repl-acement/discussions/27).

Can there be an alternative solution that would not have this problem?

**Reply by andrewzhurov** [reply]
Such an alternative solution would need to solve:
> [how share code](https://github.com/repl-acement/repl-acement/discussions/23#discussioncomment-2963284)
> [how to discover code](https://github.com/repl-acement/repl-acement/discussions/23#discussioncomment-2963353)

In order to allow users to share code of their liking they could have a personal storage they write to.
In order to discover code there can be aggregation storages from authorities.
Or, in order to discover code, there can be a web of trust, allowing for discovery of personal storages.

**Reply by raymcdermott** [reply]
what do you propose?

**Reply by andrewzhurov** [reply]
It seems there are many possible solutions our problems can be solved with.
I propose to consider possible solutions to find the best fit.
There are some I can think of, bringing more to our attention is welcome.
* in order to host code one could:
  * upload to a public server
    * i.e., a central repo
    * may have junk/spam uploads
    * requires funds to run server
    * requires trust in authority behind the server
    * authority has power over content
    * authority has power over access
  * upload to a personal server
    * authority has power over content
      * i.e., what's hosted
    * authority has power over access
      * i.e., who can read what, who can write what
    * requires hosting a personal server
  * upload to a distributed system
    * e.g.,
      * IPFS
      * Datashards
      * Tahoe LAFS
      * Freenet
      * Libchop
      * SAFE Network
        * not sure it uses content-addressing
      * FileCoin
    * noone has authority over content and access
    * fault-tolerant
* in order to make their content content discoverable one could:
  * have a mutable name that points to immutable storage
    * e.g.,
      * an IPv6 constant name
      * a DNS name
      * an IPNS name
      * Mutable Datashards capabilities
* in order to make other's content discoverable one could:
  * list known names
    * e.g.,
Alice has DNS name of https://alice.me/ that points to her latest content, and
Bob has DNS name of https://bob.me/ that points to his latest content,
resolving to e.g., {:known-names #{"https://alice.me/"}},
then Carol, knowing Bob's DNS name, would be able to access Alice's content.
  * list other's content
    * e.g.,
Alice has DNS name https://alice.me/code/latest that points her latest code on IFPS https://gateway.ipfs.io/QmUVTKsrYJpaxUT7dr9FpKq6AoKHhEM7eG1ZHGL56haKLG,
resolving to e.g., #{(fn A1 ...) (fn A2 ...)}, and
Bob has DNS name https://bob.me/code/latest that points to his latest code on IPFS https://gateway.ipfs.io/QmaVfeg2GM17RLjBs9C4fhpku6uDgrEGUYCTC183VrZaVW,
that include's Alice's code,
resolving to e.g., #{(fn A1 ...) (fn B1 ...)}.
* in order to discover content one could:
  * query for content of profiles, recursively
  * query for aggregated content of profiles

**Reply by andrewzhurov** [reply]
More concretely, one composed solution is:

Have profiles with latest content.
https://alice.me/

Allow to link from profile to profile.
```
:profile/name "https://bob.me/"
:profile/known-profile-names #{"https://alice.me/"}
```
(that makes possible querying for discovery of latest content)

(Optionally) Have a key, per profile, to sign content.
(Optionally) Have `:profile/trusted-keys` for web-of-trust.

Host content on IPFS.

Access via XTDB.

**Reply by raymcdermott** [reply]
Interesting ... how simple is it to integrate IPFS & XTDB?

**Reply by andrewzhurov** [reply]
That's a good question I don't have competence to answer.:)
Having IPFS as storage layer would be enough, it seems.
You mentioned this solution [here](https://clojurians.slack.com/archives/C03FF6W62A3/p1655131140281719?thread_ts=1655124492.445059&cid=C03FF6W62A3) and Jeremy seems to approve, so I guess at least it's possible.
I'm curious what Jeremy thinks on that.

**Reply by andrewzhurov** [reply]
> Creating a new doc-store or tx-log implementation isn't too hard, since the protocol is quite small and there are several examples to study
If you already have working Clojure + IPFS code/APIs, then I wouldn't imagine it taking more than a day or two to create a version of either
Jeremy

Examples: [S3 storage](https://github.com/xtdb/xtdb/tree/e2f51ed99fc2716faa8ad254c0b18166c937b134/modules/s3), [Datastore storage](https://github.com/avisi-apps/crux-datastore)
[Docs of storages](https://docs.xtdb.com/storage/).

**Comment by raymcdermott** [comment]
Looks pretty difficult to achieve decentralized though I agree we need to beg for resources to host the data centrally. So far, I'm hearing positive vibes about community support to sponsor this.

I'm definitely interested in how we can create a more humane / web of trust type authorization process.
